Add Ui elements and navigation for read only program view

jessbumblebee opened the pull request:
### Description

Add Ui elements and navigation for read only program view. This is the second half of the code that implements Phase 2 in the prd for improving program publishing (https://docs.google.com/document/d/1oD8EaFejs6L7_I-LKxt5ixQRFi5gtcTeL6NOwPMahf0/edit?resourcekey=0-76DVYMd4iDDbpA243q6l_Q#heading=h.atpwj1mzv74n)

## Release notes

N/A The feature is still disabled via feature flag

### Checklist

#### General

- [x] Added the correct label: < feature | bug | dependencies | infrastructure | ignore-for-release >
- [x] Created tests which fail without the change (if possible)
- [x] N/A- should be updated when the feature goes live. Extended the README / documentation, if necessary

#### User visible changes

- [x] Wrote browser tests using the [validateAccessibility](https://sourcegraph.com/github.com/civiform/civiform/-/blob/browser-test/src/support/index.ts?L437:14&subtree=true) method
- [x] Manually tested at 200% size
- [x] Manually evaluated tab order

#### New Features

- [x] Add new FeatureFlag env vars to `server/conf/application.conf`
- [x] Conditioned new functionality on a [FeatureFlag](https://docs.civiform.us/contributor-guide/developer-guide/feature-flags)
- [x] Wrote browser tests with the feature flag off and on, etc.

### Issue(s) this completes

None, because the feature isn't enabled yet

jessbumblebee commented on browser-test/src/admin_program_creation.test.ts:
```diff
@@ -19,6 +20,26 @@ describe('program creation', () => {
     await validateScreenshot(page, 'program-description-page')
   })
 
+  it('View active program, without draft and after creating draft', async () => {
```
To explain my test strategy: This test covers all the new elements. I implemented it for the new flow only, because others tests go through the existing flow as a side effect already, but do not include the clicking of the view button for example.
In other words, this was the least effort way in which both the existing and the new flow get coverage.

In an ideal world the tests would automatically switch to the flag value that is used in staging or production and pass (e.g. if the flag is submitted). Based on our eng discussion this isn't currently supported. Enabling the flag will therefore require some changes to the tests.

Let me know in case you have come across a better approach or close this comment to acknowledge that you've read it and agree.

shanemc-goog replied:
I think you did the test as expected.  The tests shouldn't know or care about any environment, they should set the state they are testing and test that.

"In an ideal world the tests would automatically switch to the flag value that is used in staging or production and pass"

This doesn't line up with what we were trying to suggest, so sorry if there was confusion.

Something you should do though that I recently discovered, is explicitly disable the flag for any tests where it being on would make them break.   there's not `disableFeatureFlag` similar to `enableFeatureFlag` for that.  This will make the tests fully compatible when they're used as "probers" for environments with the feature enabled.

The easiest way to find those tests is to turn the flag on in your local `server/conf/application.dev-browser-tests.conf` and see which other tests fail and add the disable there.

jessbumblebee commented on server/app/controllers/admin/AdminProgramBlocksController.java:
```diff
@@ -44,27 +47,59 @@ public AdminProgramBlocksController(
       ProgramService programService,
       QuestionService questionService,
       ProgramBlockEditView editView,
+      ProgramBlockEditView readOnlyView,
```
This comment should stay open until the view is replaced with a read only version (blocked unit the other pull request is submitted)

jessbumblebee commented:
Hi Shane, I've left 2 comments about a change that I will do when the other pull request is submitted (and before I submit this one) and about my testing strategy. Please take those into account when reviewing.

shanemc-goog commented:
@jessbumblebee Did you mean to request a review?  There's no reviewers marked but I see you addressed me.

Also I think you need to remove the "ignore-for-release" label, only one of the labels is meant to be set in that checklist item.

Also, you can check the checkboxes in the checklist from the rendered UI. It looks like you're manually editing it and putting an 'x' in

jessbumblebee commented on server/test/controllers/admin/AdminProgramControllerTest.java:
```diff
@@ -174,8 +174,25 @@ public void edit_returnsExpectedForm() throws Exception {
+    assertThat(result.status()).isEqualTo(OK);
     assertThat(contentAsString(result)).contains("test program");
+
+    // TODO(jhummel) uncomment when the readOnlyView is used before submitting.
```
This comment needs to be removed before submitting and after the other pull request is submitted.

jessbumblebee replied:
Remove extra blank line then too.

jessbumblebee commented:
Hi Shane, apologies for the confusion, I requested the review and then noticed I hadn't included some changes, so removed you again, added the changes and added you back. It's up to date now.

Are you saying that only one label should be selected? How do you choose if several apply?
I added the "ignore for release" because the feature will not be released with this change as it is hidden behind a flag. It is a topic I wanted to bring up in one of our team syncs because it is unclear to me how announcements work for hidden features and if this is documented anywhere.

Yes, I'm manually putting the x and placed the whitespaces wrong.

shanemc-goog commented on browser-test/src/admin_program_creation.test.ts:
```diff
@@ -19,6 +20,26 @@ describe('program creation', () => {
+    // TODO(jhummel) add screenshot here when the other pull request is submitted
+    // await validateScreenshot(page, 'program-read-only-viewer')
+    await adminPrograms.createNewVersionWithReadOnlyViewEnabled(programName)
+    await validateScreenshot(page, 'program-list-active-and-draft-program')
```
**issue (blocking):** Screenshots are great, but we should also be doing dom element checks to make the error messages more pointed towards issues.  See other tests in this file for example.

shanemc-goog commented on server/test/controllers/admin/AdminProgramBlocksControllerTest.java:
```diff
@@ -99,6 +110,74 @@ public void create_withProgram_addsRepeatedBlock() {
+            .withName("Admin name")
+            .withDescription("Admin description")
+            .build();
+    Question appName = testQuestionBank.applicantName();
```
**suggestion (non-blocking):** Use applicantName, app is too ambiguous with application

shanemc-goog commented on server/test/controllers/admin/AdminProgramBlocksControllerTest.java:
```diff
@@ -99,6 +110,74 @@ public void create_withProgram_addsRepeatedBlock() {
+        // Similar to program index page we don't show admin name.
+        .doesNotContain("Admin name");
+
+    QuestionDefinition questionDefinition =
```
**suggestion (non-blocking):** This is a bit vague as it's not clear the utility of this question.  Suggest naming something like otherQuestionDef, differentQuestionDef

shanemc-goog commented on browser-test/src/support/admin_programs.ts:
```diff
@@ -522,9 +522,29 @@ export class AdminPrograms {
+    await this.createNewVersionWithFlag(programName, true)
+  }
+
+  async createNewVersionWithFlag(
```
**issue (blocking):** `WithFlag` is not a readable or standard naming scheme, you don't need it at all since you can overload the method name and provide method docs to make it clear what the method does.

shanemc-goog commented on browser-test/src/support/admin_programs.ts:
```diff
@@ -522,9 +522,29 @@ export class AdminPrograms {
+    await this.createNewVersionWithFlag(programName, false)
+  }
+
+  async createNewVersionWithReadOnlyViewEnabled(programName: string) {
```
**suggestion (non-blocking):** You don't need this method since it calls the next. It's just as easy for callers to do

`createNewVersion('program name', /* isReadOnlyView= */ true)`

as

`createNewVersionWithReadOnlyViewEnabled('program name')`

shanemc-goog commented on server/app/controllers/admin/AdminProgramBlocksController.java:
```diff
@@ -44,27 +47,59 @@ public AdminProgramBlocksController(
+   * (blocks) if applicable through links on the page.
+   */
+  @Secure(authorizers = Authorizers.Labels.CIVIFORM_ADMIN)
+  public Result readOnlyIndex(long programId) {
```
**issue (blocking):** These should have browser tests but I don't see any.

shanemc-goog commented:
> Hi Shane, apologies for the confusion, I requested the review and then noticed I hadn't included some changes, so removed you again, added the changes and added you back. It's up to date now.
> 
> Are you saying that only one label should be selected? How do you choose if several apply? I added the "ignore for release" because the feature will not be released with this change as it is hidden behind a flag. It is a topic I wanted to bring up in one of our team syncs because it is unclear to me how announcements work for hidden features and if this is documented anywhere.

Yes, only one label from that set of labels should be used.  Ignore is correct here since this is not available yet.

These labels are for PRs and how they relate to the weekly Release Notes only, announcing new feature availability could be included in a Release Notes (in coordination with the oncall person or through the final PR for the feature), but it also should have separate communication as managed by the PM involved I would expect.  

> 
> Yes, I'm manually putting the x and placed the whitespaces wrong.

dkatzz commented on server/app/views/admin/programs/ProgramBlockBaseView.java:
```diff
@@ -94,11 +83,17 @@ protected final DivTag renderExistingPredicate(
-  protected abstract String getEditButtonUrl(ProgramDefinition programDefinition);
+  /**
+   * Returns a standardized Edit Button that can be added to the program info. A typical use case
+   * would be for a subclass to - create a button with this method - add subclass specific
```
**suggestion (blocking):** I think these are meant to be bullet points? Could you update the formatting if so?

jessbumblebee reviewed (commented):
Thanks for the comments.

Please note that I did some additional changes to tests that were not yet adaptedto changes I made after the pull request with the other half of the change was submitted.

jessbumblebee commented:
Ps: With regards to comments Shane left on the pull request, I closed the ones that were relatively trivial and have been addressed and left those open where it would be good to get follow up cofirmed by the reviewer. If you could have a look at those and close them if you support my approach, that would be great

bion commented on server/app/controllers/admin/AdminProgramBlocksController.java:
```diff
@@ -34,8 +35,11 @@
 /** Controller for admins editing screens (blocks) of a program. */
 public final class AdminProgramBlocksController extends CiviFormController {
 
+  private static final boolean READ_ONLY = true;
```
**suggestion (blocking):** With its current usage this variable doesn't look necessary

bion commented on server/app/controllers/admin/AdminProgramBlocksController.java:
```diff
@@ -128,6 +163,23 @@ public Result edit(Request request, long programId, long blockId) {
   }
 
+  /**
+   * f Returns an HTML page displaying all configurations of the specified program screen (block) as
```
**suggestion (blocking):** 
```suggestion
   * Returns an HTML page displaying all configurations of the specified program screen (block) as
```

bion commented on server/app/controllers/admin/AdminProgramController.java:
```diff
@@ -123,25 +127,35 @@ public Result publish() {
-      return redirect(controllers.admin.routes.AdminProgramBlocksController.edit(idToEdit, 1));
+      return redirect(
+          controllers.admin.routes.AdminProgramBlocksController.edit(
+              idToEdit, /* blockDefinitionId = */ 1));
```
**praise:** thanks!

bion commented on server/app/controllers/admin/NotViewableException.java:
```diff
@@ -0,0 +1,12 @@
+package controllers.admin;
+
+/**
+ * Exception when a request is sent to view something that can not be viewed/shown This class is
```
**nitpick (blocking):** 
```suggestion
 * Exception when a request is sent to view something that can not be viewed/shown. This class is
```

bion commented on server/app/repository/VersionRepository.java:
```diff
@@ -286,6 +286,11 @@ public boolean isDraftProgram(Long programId) {
         .anyMatch(draftProgram -> draftProgram.id.equals(programId));
   }
 
+  public boolean isActiveProgram(Long programId) {
```
**suggestion (blocking):** add javadoc e.g. "True if program is a member of the current active version."

bion commented on server/app/views/admin/programs/ProgramBlockBaseView.java:
```diff
@@ -94,11 +83,22 @@ protected final DivTag renderExistingPredicate(
+   * would be for a subclass to
+   *
+   * <ul>
+   *   <li><create a button with this method
```
**nitpick (blocking):** 
```suggestion
   *   <li>create a button with this method
```

bion commented on server/app/views/admin/programs/ProgramBlockBaseView.java:
```diff
@@ -94,11 +83,22 @@ protected final DivTag renderExistingPredicate(
+   *
+   * <ul>
+   *   <li><create a button with this method
+   *   <li>add subclass specific
+   *   <li>navigation behaviour to the button
```
**suggestion (blocking):** 
```suggestion
   *   <li>add subclass specific navigation behaviour to the button
```

bion commented on server/app/views/components/Icons.java:
```diff
@@ -85,6 +85,15 @@ public enum Icons {
           + " 2.792Q14 2.271 14.719 2.271Q15.438 2.271 15.958 2.792L17.208 4.042Q17.708 4.542"
           + " 17.708 5.281Q17.708 6.021 17.208 6.521L6.229 17.5ZM15.958 5.271 14.729 4.042ZM13.5"
           + " 7.729 12.896 7.104 12.271 6.5V6.479L13.5 7.729Z"),
+  VIEW(
```
**suggestion (blocking):** please add a comment either a description of the icon or linking to a visual reference of the icon

bion commented on server/test/controllers/admin/AdminProgramControllerTest.java:
```diff
@@ -174,11 +174,46 @@ public void edit_returnsExpectedForm() throws Exception {
+    assertThat(result.status()).isEqualTo(OK);
     assertThat(contentAsString(result)).contains("test program");
+
+    // TODO(jhummel) uncomment when the readOnlyView is used before submitting.
+    // assertThat(contentAsString(result)).contains("View program");
```
**issue (blocking):** remove before submitting or update comment

bion commented on server/app/views/admin/programs/ProgramBlockEditView.java:
```diff
@@ -93,10 +90,6 @@ public ProgramBlockEditView(
     this.featureFlags = checkNotNull(featureFlags);
     this.featureFlagOptionalQuestions = checkNotNull(config).hasPath("cf.optional_questions");
     this.programDisplayType = programViewType;
-
-    if (!programDisplayType.equals(DRAFT)) {
-      throw new UnsupportedOperationException(NOT_YET_IMPLEMENTED_ERROR_MESSAGE);
-    }
```
**praise:** woohoo!

bion commented on browser-test/src/admin_program_list.test.ts:
```diff
@@ -1,11 +1,17 @@
     const {page, adminPrograms} = ctx
 
     await loginAsAdmin(page)
+    await disableFeatureFlag(page, 'program_read_only_view_enabled')
```
The downside is that we have less confidence that we're not breaking critical behavior when the feature flag is turned on.

My vote would be to duplicate the tests that break when it's so there's one version with it off and one with it on, and then delete the ones with it off when we remove the flag.

bion commented on browser-test/src/admin_program_viewing.test.ts:
```diff
@@ -0,0 +1,55 @@
+describe('admin program view page', () => {
+  const ctx = createTestContext()
+
+  it('view active program, without draft and after creating draft', async () => {
```
**suggestion (blocking):** please add some questions to the program

jessbumblebee replied:
Done.

bion replied:
It looks like this test case still doesn't have a question added to the program.

bion replied:
Ah got it, sounds good

bion reviewed (changes requested):
Great work!

The questions themselves in the block view have some edit controls e.g. the optional toggle and soon the address verification toggle. Need to account for those when rendering the read only version.

dkatzz commented on browser-test/src/admin_program_viewing.test.ts:
```diff
@@ -0,0 +1,57 @@
+    const programName = 'Apc program'
+    await adminPrograms.addProgram(programName)
+    await adminPrograms.publishAllPrograms()
+    await adminPrograms.gotoViewActiveProgramPage(programName)
```
Sounds good, thanks for the clarification!

bion commented on server/app/controllers/admin/AdminProgramBlocksController.java:
```diff
@@ -44,28 +46,59 @@ public final class AdminProgramBlocksController extends CiviFormController {
     this.formFactory = checkNotNull(formFactory);
     this.requestChecker = checkNotNull(requestChecker);
   }
 
   /**
-   * Return a HTML page displaying all configurations of the specified program.
+   * Returns an HTML page displaying all configurations of the specified program and UI elements to
+   * start editing aspects of the program.
```
**nitpick:** I'm not sure if this javadoc is correct. I think this page shows the edit page for the last block of the program?

bion replied:
Thank you for the clear explanation!

bion commented on server/app/controllers/admin/AdminProgramBlocksController.java:
```diff
@@ -44,28 +46,59 @@ public final class AdminProgramBlocksController extends CiviFormController {
+  }
+
+  /**
+   * Returns an HTML page displaying a read only version of all configurations of the specified
+   * program.
```
**suggestion:** 
```suggestion
   * Returns an HTML page displaying a read only version of the program's block configuration.```

bion commented on server/conf/routes:
```diff
@@ -17,6 +17,7 @@ GET     /securePlayIndex            controllers.HomeController.securePlayIndex()
 # A controller for pages for an admin to create and maintain programs
 GET     /admin/programs                                         controllers.admin.AdminProgramController.index(request: Request)
 GET     /admin/programs/new                                     controllers.admin.AdminProgramController.newOne(request: Request)
+GET     /admin/programs/:programId/view                         controllers.admin.AdminProgramController.view(request: Request, programId: Long)
```
**suggestion (blocking):** rename `view` to `show` here and in the controller, following the [routing convention](https://docs.civiform.us/contributor-guide/developer-guide/development-standards?q=standards#html-routing-convention)

bion commented on server/app/repository/VersionRepository.java:
```diff
@@ -277,15 +277,25 @@ public boolean isDraft(Question question) {
   }
 
+  /**
+   * Returns true if the program with the provided id is a member of the current active versions.
```
**praise:** 
```suggestion
   * Returns true if the program with the provided id is a member of the current active version.
```

bion replied:
Singular because `Version` is a specific resource in CiviForm and there is only one active Version at a time

jessbumblebee reviewed (commented):
Heads up: I have closed Shane's comments as nobody added additional comments to them in the last round of reviews

dkatzz reviewed (approved):
A couple of small nits but overall LGTM!
